libraries/AP_GyroFFT/AP_GyroFFT.cpp

```diff
@@ -210,122 +242,120 @@ void AP_GyroFFT::init(uint32_t target_looptime_us)
-    _samples_per_frame = 1 << lrintf(log2f(_samples_per_frame));
-
-    // The update rate for the output
+    // per-axis frame time
+    _frame_time_ms = _samples_per_frame * 1000 / _fft_sampling_rate_hz;
+    // The update rate for the output, defaults are 1Khz / (1 - 0.5) * 32 == 62hz
     const float output_rate = _fft_sampling_rate_hz / _samples_per_frame;
     // establish suitable defaults for the detected values
     for (uint8_t axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
         _thread_state._center_freq_hz[axis] = _fft_min_hz;
-        _thread_state._center_freq_hz_filtered[axis] = _fft_min_hz;
-        // calculate low-pass filter characteristics based on overlap size
-        _center_freq_filter[axis].set_cutoff_frequency(output_rate, output_rate * 0.48f);
-        // smooth the bandwidth output more aggressively
-        _center_bandwidth_filter[axis].set_cutoff_frequency(output_rate, output_rate * 0.25f);
+
+        for (uint8_t peak = 0; peak < FrequencyPeak::MAX_TRACKED_PEAKS; peak++) {
+            _thread_state._center_freq_hz_filtered[axis][peak] = _fft_min_hz;
+        }
         // number of cycles to average over, two complete windows to be sure
         _noise_calibration_cycles[axis] = (_window_size / _samples_per_frame) * 2;
+        // harmonic frequency fit should change relatively slowly
+        _harmonic_fit_filter[axis].set_cutoff_frequency(output_rate, MIN(output_rate * 0.48f, FFT_HARMONIC_FIT_FILTER_HZ));
+    }
+
+    // configure a filter for frequency, bandwidth and energy for each of the three tracked noise peaks
+    for (uint8_t peak = 0; peak < FrequencyPeak::MAX_TRACKED_PEAKS; peak++) {
+        // calculate low-pass filter characteristics based on window size and overlap
+        _center_freq_filter[peak].set_cutoff_frequency(output_rate, output_rate * 0.48f);
+        // the bin energy jumps around a lot so requires more filtering
+        _center_freq_energy_filter[peak].set_cutoff_frequency(output_rate, output_rate * 0.25f);
+        // smooth the bandwidth output more aggressively
+        _center_bandwidth_filter[peak].set_cutoff_frequency(output_rate, output_rate * 0.25f);
     }
 
     // the number of cycles required to have a proper noise reference
     _noise_cycles = (_window_size / _samples_per_frame) * XYZ_AXIS_COUNT;
-    // calculate harmonic multiplier
-    uint8_t first_harmonic = 0;
-    _harmonics = 1; // always search for 1
-    for (uint8_t i = 0; i < HNF_MAX_HARMONICS; i++) {
-        if (_ins->get_gyro_harmonic_notch_harmonics() & (1<<i)) {
-            if (first_harmonic == 0) {
-                first_harmonic = i + 1;
-            } else {
-                _harmonics++;
-                _harmonic_multiplier = float(i + 1) / first_harmonic;
-                break;
-            }
-        }
-    }
 
     // finally we are done
     _initialized = true;
     update_parameters();
+    set_analysis_enabled(true);
     // start running FFTs
     start_update_thread();
```

Comment:
maybe enable analysis once we know the thread can be started too? could be first thing the thread does, or make start_update_thread return bool

Reply:
Done